### PR TITLE
[Cloud Deployment IVa] EFS creation and mounting

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
breaks up various parts of #393, starting with EFS handling

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def read_requirements(file):
 
 
 extras_require = defaultdict(list)
-extras_require["full"] = ["dandi>=0.58.1", "hdf5plugin"]
+extras_require["full"] = ["dandi>=0.58.1", "hdf5plugin", "boto3"]
```
Comment (CodyCBakerPhD) [inline setup.py:18]: Looks like this got missed in the pyproject.toml refactor

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 2]
```diff
@@ -131,6 +141,29 @@ def submit_aws_batch_job(
         job_queue_name=job_queue_name, compute_environment_name=compute_environment_name, batch_client=batch_client
     )
 
+    # Create or fetch EFS volume and attach it to the job
+    efs_id = None
```
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:145]: Will similarly put this in a private helper later, to match style of other steps

File: tests/test_minimal/test_tools/aws_tools.py [hunk 3]
```diff
@@ -1,10 +1,13 @@
+import datetime
 import os
 import time
 
 import boto3
 
 from neuroconv.tools.aws import submit_aws_batch_job
 
+_RETRY_STATES = ["RUNNABLE", "PENDING", "STARTING", "RUNNING"]
```
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/aws_tools.py:9]: This is one of those bug fixes I mentioned sneaking into this PR - after running the tests so many times I did see some edge cases show up in the waiting times

I think a final follow-up in the series (once everything works together) could be more elegant with some common utils for such things

File: tests/test_minimal/test_tools/aws_tools.py [hunk 4]
```diff
@@ -35,14 +38,24 @@ def test_submit_aws_batch_job():
...
     job_id = info["job_submission_info"]["jobId"]
+    job = None
+    max_retries = 10
+    retry = 0
+    while retry < max_retries:
+        job_description_response = batch_client.describe_jobs(jobs=[job_id])
+        assert job_description_response["ResponseMetadata"]["HTTPStatusCode"] == 200
+
+        jobs = job_description_response["jobs"]
+        assert len(jobs) == 1
```
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/aws_tools.py:49]: This is a repeated example of the wait condition until a desired outcome is reached

File: tests/test_minimal/test_tools/aws_tools.py [hunk 5]
```diff
@@ -156,3 +180,120 @@ def test_submit_aws_batch_job_with_dependencies():
...
+    )
+
+    # Cleanup EFS after testing is complete - must clear mount targets first, then wait before deleting the volume
+    # TODO: cleanup job definitions? (since built daily)
+    mount_targets = efs_client.describe_mount_targets(FileSystemId=efs_id)
+    for mount_target in mount_targets["MountTargets"]:
+        efs_client.delete_mount_target(MountTargetId=mount_target["MountTargetId"])
+
+    time.sleep(60)
+    efs_client.delete_file_system(FileSystemId=efs_id)
```
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/aws_tools.py:295]: The the future I plan to be more elegant with pytest cleanup, possibly via `pytest_sessionfinish` in a `conftest.py`

It might also cleanup any job definitions that start with 'test_neuroconv'

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 6]
```diff
@@ -534,6 +627,20 @@ def _ensure_job_definition_exists_and_get_arn(
     minimum_time_to_kill_in_days = 1  # Note: eventually consider exposing this for very long jobs?
     minimum_time_to_kill_in_seconds = minimum_time_to_kill_in_days * 24 * 60 * 60
 
+    volumes = []
+    mountPoints = []
+    if efs_id is not None:
+        volumes = [
+            {
+                "name": "neuroconv_batch_efs_mounted",
```
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:635]: The name identifier here is entirely local (does not need to match the 'name' tagged on the actual EFS volume)

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 7]
```diff
@@ -449,6 +536,8 @@ def _generate_job_definition_name(
     job_definition_name += f"_{parsed_docker_image_name}-image"
     job_definition_name += f"_{minimum_worker_ram_in_gib}-GiB-RAM"
     job_definition_name += f"_{minimum_worker_cpus}-CPU"
+    if efs_id is not None:
+        job_definition_name += f"_{efs_id}"
```
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:540]: Turned out to be quite important to include the filesystem ID in the definition name to avoid reusing a previous definition created without the EFS mount configured

I have toyed with the idea of making the job definition a hash of all the 'unique' configuration aspects, but it is also important for it to be readable - a follow-up towards the end might apply a bunch of human readable tags to this effect

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 8]
```diff
@@ -305,8 +321,8 @@ def _ensure_compute_environment_exists(
             "type": "EC2",
             "allocationStrategy": "BEST_FIT",  # Note: not currently supporting spot due to interruptibility
             "instanceTypes": ["optimal"],
-            "minvCpus": 1,
-            "maxvCpus": 8,  # Not: not currently exposing control over this since these are mostly I/O intensive
+            "minvCpus": 0,  # Note: if not zero, will always keep an instance running in active state on standby
+            "maxvCpus": 8,  # Note: not currently exposing control over this since these are mostly I/O intensive
```
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:325]: @h-mayorquin This is a very important detail to be aware of

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 9]
```diff
@@ -391,9 +407,20 @@ def _ensure_job_queue_exists(
...
         jobStateTimeLimitActions=[
             dict(
-                reason="Avoid zombie jobs.",
+                reason="MISCONFIGURATION:COMPUTE_ENVIRONMENT_MAX_RESOURCE",
+                state="RUNNABLE",
+                maxTimeSeconds=minimum_time_to_kill_in_seconds,
+                action="CANCEL",
+            ),
+            dict(
+                reason="MISCONFIGURATION:JOB_RESOURCE_REQUIREMENT",
```
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:423]: Another bug fix I'm sneaking in here - previous testing suites ran under an existing compute environment that did not have this, but now it is actually setup properly (and tested in practice but no clue how to make a proper test for it) and will end any zombie jobs

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin This is ready for review now

Evidence of passing tests in CI: https://github.com/catalystneuro/neuroconv/actions/runs/10726561768/job/29746896174

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py [hunk 10]
```diff
@@ -42,6 +43,9 @@ def submit_aws_batch_job(
         E.g., `commands=["echo", "'Hello, World!'"]`.
     environment_variables : dict, optional
         A dictionary of environment variables to pass to the Docker container.
+    efs_volume_name : str
```
Comment (h-mayorquin) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:46]: docstring should say optional.

**Review by h-mayorquin — APPROVED** [review]
LGTM after discussion.

File: src/neuroconv/tools/aws/_submit_aws_batch_job.py (hunk 10 above)
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/aws/_submit_aws_batch_job.py:46]:
```suggestion
    efs_volume_name : str, optional
```
